### ISSUE TITLE
Placeholder/argument format

### ISSUE THREAD

**Issue by StefanHaug** [JeremySkinner/FluentValidation#905]
I see there's quite a few questions regarding formatting of the placeholders.

I must say that would be a nice addition to the framework. We use templates for misc. error messages and if we decide change the date/time representation of a certain message, now we have to change it all the places we use it instead of the template itself.

I know you've said you've no immediate plans for this feature, but I just wanted to give you another reason why this would be useful. Separation of construction and representation as it were.

**Comment by JeremySkinner** [comment]
The placeholder system is designed to only be a simple mechanism for doing simple string replacement for common use cases. It's really not intended to be a full blown alternative to string.Format. 

If you need this kind of flexibility then you should really be using using the overloads of `WithMessage` that take a lambda and directly call string.Format on your template inside the callback.

**Comment by StefanHaug** [comment]
I might be able to introduce this as a pull request if you're comfortable with having this feature.

**Comment by JeremySkinner** [comment]
Potentially yes, could you provide some details on what you're planning on implementing before you do?

**Comment by StefanHaug** [comment]
Basically being able to specify a format in the template e.g. {PropertyValue:g} which would result in string.format("{0:g}", propertyValue).

It seems the change will be very limited and won't break any tests.

**Comment by JeremySkinner** [comment]
Ok great, yes please feel free to send a PR and I’ll be happy to take a look.

**Comment by StefanHaug** [comment]
I've not collaborated much using GIT, so I was unaware that I had to create a new branch. Trying to revert/undo the commits (which failed to push) was beyond me, so I'll paste my code here.

New method in MessageFormatter.cs:

		// TODO: Add alignment component
		private Regex _keyRegex = new Regex("{([^{}:]+)(?::([^{}]+))?}"); // Matches any placeholder

		protected virtual string ReplacePlaceholdersWithValues(string template, IDictionary<string, object> values)
		{
			return _keyRegex.Replace(template, m =>
			{
				var key = m.Groups[1].Value;

				if (!values.ContainsKey(key))
					return m.Value; // No placeholder / value

				var format = m.Groups[2].Success // Format specified?
					? $"{{0:{m.Groups[2].Value}}}"
					: null;

				return format == null
					? values[key]?.ToString()
					: string.Format(format, values[key]);
			});
		}

Replace assignment of `result` in `public virtual string BuildMessage(string messageTemplate)` to:

			result = ReplacePlaceholdersWithValues(result, _placeholderValues);

Unit tests:

		[Fact]
		public void Format_property_value()
		{
			using (new TemporalCultureExchange())
			{
				string result = formatter
				.AppendPropertyValue(123.45)
				.BuildMessage("{PropertyValue:#.#}");

				result.ShouldEqual("123.5");
			}
		}

		[Fact]
		public void Understands_numeric_formats()
		{
			using (new TemporalCultureExchange())
			{
				string result = formatter
					.AppendArgument("d", 123)
					.AppendArgument("e", 1052.0329112756)
					.AppendArgument("c", -123.456789)
					.AppendArgument("f", 1234.567)
					.AppendArgument("p", 0.912)
					.BuildMessage("{c:c3} {c:c4} {f:f} {p:p} {d:d} {e:e} {e:0.00} {d:0000}");

				result.ShouldEqual("($123.457) ($123.4568) 1234.57 91.20% 123 1.052033e+003 1052.03 0123");
			}
		}

		[Fact]
		public void Understands_date_formats()
		{
			var now = DateTime.Now;
			using (new TemporalCultureExchange())
			{
				string result = formatter
					.AppendArgument("now", now)
					.BuildMessage("{now:g} {now:MM-dd-yy} {now:f}");

				result.ShouldEqual($"{now.ToString("g")} {now.ToString("MM-dd-yy")} {now.ToString("f")}");
			}
		}

		[Fact]
		public void Should_ignore_unknown_parameters()
		{
			var now = DateTime.Now;
			string result = formatter
				.AppendArgument("foo", now)
				.BuildMessage("{foo:g} {unknown} {unknown:format}");

			result.ShouldEqual($"{(now.ToString("g"))} {{unknown}} {{unknown:format}}");
		}

		internal class TemporalCultureExchange : IDisposable
		{
			private bool disposedValue = false;
			private CultureInfo _culture;

			internal TemporalCultureExchange(string culture = "en-US")
			{
				_culture = Thread.CurrentThread.CurrentCulture;

				Thread.CurrentThread.CurrentCulture = CultureInfo.GetCultureInfo(culture);
			}

			protected virtual void Dispose(bool disposing)
			{
				if (!disposedValue)
				{
					if (disposing)
						Thread.CurrentThread.CurrentUICulture = _culture;

					disposedValue = true;
				}
			}

			public void Dispose()
			{
				Dispose(true);
			}
		}

All tests green.

**Comment by JeremySkinner** [comment]
If you could try again to create a pull requst that'd be great. If you use a pull request then it'll automatically run the tests and will also allow me to start a code review which I can't do easily here.

**Comment by StefanHaug** [comment]
Any idea how i remove the outgoing commits?

**Comment by JeremySkinner** [comment]
Not sure what you mean. The steps should be as follows:
- Fork the FluentValidation repository
- Clone your fork locally
- Make the changes and commit them
- Push to your fork
- Use the create pull request button

**Comment by JeremySkinner** [comment]
If you can't get that work, let me know what issue you're running into and I can try and walk you through it.

**Comment by StefanHaug** [comment]
I'll try, I just haven't got time at the moment. Should be within a day or two.

**Comment by JeremySkinner** [comment]
Thanks. I'll do a full review once you've sent that through.

**Comment by StefanHaug** [comment]
I've created the pull request, but it seemed to fail.

**Comment by JeremySkinner** [comment]
Thanks - yes the new tests are failing - could you take a look at that? I've left a code review on the PR with some comments too.

**Comment by StefanHaug** [comment]
What's the correct procedure when updating? I just commit to the fork again and issue a new pull request?

**Comment by JeremySkinner** [comment]
No need to open a new pull request, just commit to the same fork and the pull request will be updated

**Comment by jzabroski** [comment]
@StefanHaug If you want to do a screen share, I can walk you through it as well.

**Comment by StefanHaug** [comment]
No, that's ok. It seems straight forward enough. Thanks anyways @jzabroski , I appreciate it! I'll get around to it, i'm just short of time.
